### ISSUE TITLE
Incorrect confidence output after loading a model

### ISSUE THREAD

**Issue by jrmarkle** [VowpalWabbit/vowpal_wabbit#1806]
#### Describe the bug
Predictions are outputting `inf` confidence after loading a trained model.

#### To Reproduce
Train a logistic regression model with `--loss_function=logistic --link=logistic --confidence --save_resume`. Save it, load it and make a prediction.

See my latest [example app](https://github.com/jrmarkle/vw-memory-test) with a `Dockerfile` that shows the problem. (Comment out the `patch` command to see the problem.)

#### Expected behavior
Saving and loading the model should not change the confidence output.

#### Observed Behavior
The loaded model outputs `inf` confidence.

#### Environment
Latest `master` on Ubuntu 18.04.

#### Additional context
I have narrowed down the change where this started happening to a single line in commit e833199eb89a752524159a4a1f7145d4357e63e8. In `gd.cc` the last parameter in the call to `compute_rate_decay` was changed from `fw` to `w[0]`. See [revert-e833199e.patch](https://github.com/jrmarkle/vw-memory-test/blob/master/revert-e833199e.patch). If I apply that patch to the latest commit on `master` the tests fail but the confidence output is the same before and after loading.

Using the latest `master` the output of my [example app](https://github.com/jrmarkle/vw-memory-test) is
```
predict good = 0.996043, confidence = 12.814723
predict bad = 0.004103, confidence = 13.420224
predict unknown = 0.475109, confidence = 0.000000
-------
predict good = 0.996043, confidence = inf
predict bad = 0.004103, confidence = inf
predict unknown = 0.475109, confidence = inf
```
When I apply [revert-e833199e.patch](https://github.com/jrmarkle/vw-memory-test/blob/master/revert-e833199e.patch) the output is
```
predict good = 0.996043, confidence = 12.814723
predict bad = 0.004103, confidence = 13.420224
predict unknown = 0.475109, confidence = 0.000000
-------
predict good = 0.996043, confidence = 12.814723
predict bad = 0.004103, confidence = 13.420224
predict unknown = 0.475109, confidence = 0.000000
```

**Comment by pavithraks** [comment]
Hey @jrmarkle  , apologies for the delay, I finally got around to looking at it. Thank you for identifying this.

In the example that @jrmarkle shares, the example that we pass has no label, therefore, the value of `grad_squared` (line 509 in gd.cc) is infinity. This value is coming from the function call defined in line 549 in gd.cc 
`grad_squared *= all.loss->getSquareGrad(ec.pred.scalar, ld.label);`
and defined in line 66 in `loss_functions.cc`
`float getSquareGrad(float prediction, float label) { return 4.f * (prediction - label) * (prediction - label); }`

In gd.cc line 533, originally, when we had `fw` being passed to `compute_rate_decay`, we had no problem because fw was never modified before the function call. As @jrmarkle  points out, this issue was caused because of the changes to gd.cc in this commit - e[833199](https://github.com/VowpalWabbit/vowpal_wabbit/commit/e833199eb89a752524159a4a1f7145d4357e63e8)

With this change, we now have:
`w[spare] = compute_rate_decay<sqrt_rate, adaptive, normalized>(nd.pd, w[0])`

This is problematic, because, as I mentioned earlier, the w vector is modified in line 509 as:
`w[adaptive] += nd.grad_squared * x2;`
and as indicated above, nd.grad_squared is inf.

@JohnLangford - I am not sure what the right fix for this is. Changing it back to fw causes the tests to fail. Maybe the right thing to do is to initialize "label" suitably when there is no label at all.

Thanks a ton @lokitoth for helping me navigate this. I owe you many doughnuts for this!

(The tests are failing locally irrespective of this change, which @lokitoth also confirmed. I will probably file a separate issue for that)

**Comment by pavithraks** [comment]
@JohnLangford : ^ Thoughts on how to fix this? Thanks!

**Comment by JohnLangford** [comment]
What change is causing the saving and loading of a file (even with --save_resume) to modify the prediction?  It's not the _code_ because that is identical before and after loading a file.

**Comment by JohnLangford** [comment]
Starting from the patch, "stateless" is on leading to w pointing to a different chunk of memory here:
https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/vowpalwabbit/gd.cc#L506
The core reason why "stateless" is on, is because evaluating the confidence should not change the state of the learning algorithm.

The semantics of confidence here is "how much update is required to cause the prediction to pass a threshold?" as per: https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/vowpalwabbit/confidence.cc#L30
Given this, it's not entirely defined when no label information is present, since no label -> no gradient -> no update.   What should be done? 

1) We could just document this, although it seems likely people will find it hard to use.
2) We could automatically plug in the "other" label (either -1 or 1, whichever is opposite the prediction).  Maybe this is the better solution?

**Comment by JohnLangford** [comment]
A proposed fix along the lines of 2 in #1901
